Address some of the Safer CPP warnings in DNSResolveQueueCFNet.cpp

cdumez opened the pull request:
#### ff08cabb2102fb7af646608a942999bd03f94e44
<pre>
Address some of the Safer CPP warnings in DNSResolveQueueCFNet.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=287469">https://bugs.webkit.org/show_bug.cgi?id=287469</a>

Reviewed by Darin Adler.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/posix/SocketPOSIX.h: Added.
(WTF::isIPV4SocketAddress):
(WTF::dynamicCastToIPV4SocketAddress):
(WTF::asIPV4SocketAddress):
(WTF::isIPV6Socket):
(WTF::dynamicCastToIPV6SocketAddress):
(WTF::asIPV6SocketAddress):
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::extractIPAddress):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
(WebKit::gatherNetworkMap):
(WebKit::connectToRemoteAddress):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
(WebKit::isNat64IPAddress):
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::ipAddress):
* Source/WebKit/Shared/RTCNetwork.cpp:
(WebKit::RTC::Network::IPAddress::IPAddress):

Canonical link: <a href="https://commits.webkit.org/290259@main">https://commits.webkit.org/290259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b8862f877a70a305f7ccdddf57e1a2dfed74ae7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89458 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8984 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94449 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40225 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91510 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9372 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17264 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/26572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92459 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7192 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81190 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49277 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/6927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39331 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/82256 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36574 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96278 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/88233 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16639 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77785 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16895 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76991 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77099 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19024 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21525 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9790 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16653 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/110726 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16394 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26545 "Found 5 new JSC stress test failures: microbenchmarks/memcpy-wasm-small.js.bytecode-cache, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager, wasm.yaml/wasm/stress/repro_1289.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19845 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18175 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/posix/SocketPosix.h:
```diff
@@ -0,0 +1,100 @@
+{
+    zeroBytes(storage);
+    storage.ss_family = AF_INET;
+    SUPPRESS_MEMORY_UNSAFE_CAST return initialize(*reinterpret_cast<struct sockaddr_in*>(&storage));
```
It's driving me crazy that `initialize` can fail but still leave the AF_INET type tag in place.

How about initializing ss_family to AF_UNSPEC, returning early if `initialize` fails, and only assigning AF_INET after `initialize` succeeds?

geoffreygaren commented on Source/WTF/wtf/posix/SocketPosix.h:
```diff
@@ -0,0 +1,100 @@
+inline bool initializeIPV6StocketAddress(sockaddr_storage& storage, NOESCAPE const Function<bool(sockaddr_in6&)>& initialize)
+{
+    zeroBytes(storage);
+    storage.ss_family = AF_INET6;
```
Ditto.

darinadler commented on Source/WTF/wtf/posix/SocketPosix.h:
```diff
@@ -0,0 +1,106 @@
+    SUPPRESS_MEMORY_UNSAFE_CAST return *reinterpret_cast<const struct sockaddr_in*>(&socket);
+}
+
+inline bool initializeIPV4StocketAddress(sockaddr_storage& storage, NOESCAPE const Function<bool(sockaddr_in&)>& initialize)
```
```suggestion
inline bool initializeIPV4SocketAddress(sockaddr_storage& storage, NOESCAPE const Function<bool(sockaddr_in&)>& initialize)
```

darinadler commented on Source/WTF/wtf/posix/SocketPosix.h:
```diff
@@ -0,0 +1,106 @@
+    SUPPRESS_MEMORY_UNSAFE_CAST return *reinterpret_cast<const struct sockaddr_in6*>(&socket);
+}
+
+inline bool initializeIPV6StocketAddress(sockaddr_storage& storage, NOESCAPE const Function<bool(sockaddr_in6&)>& initialize)
```
```suggestion
inline bool initializeIPV6SocketAddress(sockaddr_storage& storage, NOESCAPE const Function<bool(sockaddr_in6&)>& initialize)
```

darinadler commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -39,6 +39,7 @@
 #include <wtf/StdLibExtras.h>
 #include <wtf/URL.h>
 #include <wtf/cf/VectorCF.h>
+#include <wtf/posix/SocketPosix.h>
```
```suggestion
#include <wtf/posix/SocketPOSIX.h>
```
Consider leaving POSIX all capitals.

darinadler reviewed (approved):
A little surprised you didn’t use `is<sockaddr_in>`, `dynamicDowncast<sockaddr_in>`, `downcast<sockaddr_in>`, `is<sockaddr_in6>`, `dynamicDowncast<sockaddr_in6>`, and `downcast<sockaddr_in6>` instead of `isIPV4SocketAddress`, `dynamicCastToIPV4SocketAddress`, `asIPV4SocketAddress`, `isIPV6SocketAddress`, `dynamicCastToIPV6SocketAddress`, and `asIPV6SocketAddress`.

cdumez committed:
Address some of the Safer CPP warnings in DNSResolveQueueCFNet.cpp
https://bugs.webkit.org/show_bug.cgi?id=287469

Reviewed by Darin Adler.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/posix/SocketPOSIX.h: Added.
(WTF::isIPV4SocketAddress):
(WTF::dynamicCastToIPV4SocketAddress):
(WTF::asIPV4SocketAddress):
(WTF::isIPV6Socket):
(WTF::dynamicCastToIPV6SocketAddress):
(WTF::asIPV6SocketAddress):
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::extractIPAddress):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCMonitor.cpp:
(WebKit::gatherNetworkMap):
(WebKit::connectToRemoteAddress):
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
(WebKit::isNat64IPAddress):
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::ipAddress):
* Source/WebKit/Shared/RTCNetwork.cpp:
(WebKit::RTC::Network::IPAddress::IPAddress):

Canonical link: https://commits.webkit.org/290259@main
